hefquin-engine/src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBindJoinWithVALUESorFILTER.java

```diff
@@ -0,0 +1,91 @@
+package se.liu.ida.hefquin.engine.queryplan.physical.impl;
+
+import se.liu.ida.hefquin.base.query.TriplePattern;
+import se.liu.ida.hefquin.base.queryplan.ExpectedVariables;
+import se.liu.ida.hefquin.engine.federation.FederationMember;
+import se.liu.ida.hefquin.engine.federation.SPARQLEndpoint;
+import se.liu.ida.hefquin.engine.queryplan.executable.UnaryExecutableOp;
+import se.liu.ida.hefquin.engine.queryplan.executable.impl.ops.ExecOpBindJoinSPARQLwithVALUESorFILTER;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpTPAdd;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpTPOptAdd;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalPlanVisitor;
+
+/**
+* A physical operator that implements (a batching version of) the bind
+ * join algorithm. It starts by using a VALUES clause to capture the potential join
+ * partners that are sent to the federation member. If this fails, it uses the bind join
+ * algorithm with a FILTER clause instead.
+
+ * <p>
+ * <b>Algorithm description:</b> For a detailed description of the
+ * actual algorithm associated with this physical operator, refer
+ * to {@link ExecOpBindJoinSPARQLwithVALUESorFILTER}, which provides the
+ * implementation of this algorithm.
+ * </p>
+ */
+public class PhysicalOpBindJoinWithVALUESorFILTER extends BaseForPhysicalOpSingleInputJoin
+{
+
+	public PhysicalOpBindJoinWithVALUESorFILTER( final LogicalOpTPAdd lop ) {
+		super(lop);
+		assert lop.getFederationMember() instanceof SPARQLEndpoint;
+	}
+
+	public PhysicalOpBindJoinWithVALUESorFILTER( final LogicalOpTPOptAdd lop ) {
+		super(lop);
+
+		assert lop.getFederationMember() instanceof SPARQLEndpoint;
+	}
+
+	@Override
+	public boolean equals( final Object o ) {
+		return o instanceof PhysicalOpBindJoinWithVALUESorFILTER
+				&& ((PhysicalOpBindJoinWithVALUESorFILTER) o).lop.equals(lop);
+	}
+
+	@Override
+	public UnaryExecutableOp createExecOp( final boolean collectExceptions,
+	                                       final ExpectedVariables... inputVars ) {
+		final TriplePattern pt;
+		final FederationMember fm;
+		final boolean useOuterJoinSemantics;
+
+		if ( lop instanceof LogicalOpTPAdd ) {
+			pt = ( (LogicalOpTPAdd) lop ).getTP();
+			fm = ( (LogicalOpTPAdd) lop ).getFederationMember();
+			useOuterJoinSemantics = false;
+		}
+		else if ( lop instanceof LogicalOpTPOptAdd ) {
+			pt = ( (LogicalOpTPOptAdd) lop ).getTP();
+			fm = ( (LogicalOpTPOptAdd) lop ).getFederationMember();
+			useOuterJoinSemantics = true;
+		}
+		else {
+			throw new IllegalArgumentException("Unsupported type of operator: " + lop.getClass().getName() );
+		}
+
+		return createExecOp(pt, fm, useOuterJoinSemantics, collectExceptions);
```

Comment:
Related to my other comment, this part needs to be changed as well:
```suggestion

		if ( lop instanceof LogicalOpTPAdd ) {
			pt = ( (LogicalOpTPAdd) lop ).getTP();
			fm = ( (LogicalOpTPAdd) lop ).getFederationMember();
		}
		else {
			throw new IllegalArgumentException("Unsupported type of operator: " + lop.getClass().getName() );
		}

		return createExecOp(pt, fm, collectExceptions);
```